lib/winappdriver.js

```diff
@@ -1,150 +1,191 @@
     }
 
-    this.changeState(WinAppDriver.STATE_STARTING);
-
-    // XXXYD TODO: would be better if WinAppDriver didn't require passing in /wd/hub as a param
-    let args = [this.proxyPort + '/wd/hub'];
-
-    const startDetector = (stdout) => stdout.includes('listening for requests');
-
-    let processIsAlive = false;
-    try {
-      await this.killAll();
-
-      // set up our subprocess object
-      this.proc = new SubProcess(WAD_INSTALL_PATH, args, {
-        encoding: 'ucs2'
+    if (!this.port) {
+      await PORT_ALLOCATION_GUARD(async () => {
+        const [startPort, endPort] = WAD_PORT_RANGE;
+        try {
+          this.port = await findAPortNotInUse(startPort, endPort);
+        } catch (e) {
+          log.errorAndThrow(
+            `Could not find any free port in range ${startPort}..${endPort}. ` +
+            `Please check your system firewall settings or set 'systemPort' capability ` +
+            `to the desired port number`);
+        }
       });
-      processIsAlive = true;
-
-      // handle log output
-      for (let stream of ['STDOUT', 'STDERR']) {
-        this.proc.on(`lines-${stream.toLowerCase()}`, (lines) => {
-          for (let l of lines) {
-            log.info(`[${stream}] ${l.trim()}`);
-          }
-        });
-      }
+    }
 
-      // handle out-of-bound exit by simply emitting a stopped state
-      this.proc.on('exit', (code, signal) => {
-        processIsAlive = false;
-        if (this.state !== WinAppDriver.STATE_STOPPED &&
-            this.state !== WinAppDriver.STATE_STOPPING) {
-          let msg = `WinAppDriver exited unexpectedly with code ${code}, ` +
-                    `signal ${signal}`;
-          log.error(msg);
-          this.changeState(WinAppDriver.STATE_STOPPED);
+    const args = [`${this.port}${this.base}`];
+    // set up our subprocess object
+    this.proc = new SubProcess(WAD_INSTALL_PATH, args, {
+      encoding: 'ucs2'
+    });
+    for (const stream of ['STDOUT', 'STDERR']) {
+      this.proc.on(`lines-${stream.toLowerCase()}`, (lines) => {
+        for (const line of lines) {
+          log.info(`[${stream}] ${line.trim()}`);
         }
       });
-      log.info(`Spawning WinAppDriver with: ${args.join(' ')}`);
-
-      // start subproc and wait for startDetector
-      await this.proc.start(startDetector);
-      await this.waitForOnline();
+    }
+    this.proc.on('exit', (code, signal) => {
+      log.info(`WinAppDriver exited with code ${code}, signal ${signal}`);
+    });
+    log.info(`Spawning WinAppDriver with: ${args.join(' ')}`);
+    await this.proc.start(0);
+  }
 
-    } catch (e) {
-      this.emit(WinAppDriver.EVENT_ERROR, e);
-      // just because we had an error doesn't mean the WinAppDriver process
-      // finished; we should clean up if necessary
-      if (processIsAlive) {
-        await this.proc.stop();
-      }
-      log.errorAndThrow(e);
+  async stop () {
+    if (this.isRunning) {
+      await this.proc.stop('SIGTERM');
```

Comment:
It seems SIGINT, SIGKILL and SIGTERM have no differences on Windows...?
https://nodejs.org/docs/latest-v10.x/api/process.html

Reply:
yep, makes sense